### PR TITLE
Add S3 as storage replacement

### LINKED ISSUE #10: Add S3 support for the mod archive

### PR DESCRIPTION
**Pull request by paulhobbel**
Resolves #10

### CODE REVIEW

File: VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public string Author { get; set; }
+        public string Version { get; set; }
+        public string Tag { get; set; }
+        public Dictionary<string, string> Dependencies { get; set; }
+
+        public ModVersionArchive Archive { get; set; }
+
```
Comment (3ti65) [inline VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs:21]: remove newline

File: VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+namespace VUModManagerRegistry.Models.Dtos
+{
+    public class ModVersionArchive
+    {
+        public string ContentType { get; set; }
+        public string Data { get; set; }
+        public int Length { get; set; }
+    }
+    public class CreateModVersionRequest
```
Comment (3ti65) [inline VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs:11]: add newline

File: VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿using System.Collections.Generic;
+
+namespace VUModManagerRegistry.Models.Dtos
+{
+    public class ModVersionArchive
```
Comment (3ti65) [inline VUModManagerRegistry/Models/Dtos/CreateModVersionDto.cs:5]: these would be records in > c# 8.0

File: VUModManagerRegistry/Services/S3/S3ModStorage.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    
+    [ExcludeFromCodeCoverage]
+    public class S3ModStorage : IModStorage
+    {
+        private readonly string _bucketName;
+        private readonly AmazonS3Client _s3Client;
+
+        public S3ModStorage(IOptions<S3ModStorageOptions> options)
+        {
+            _bucketName = options.Value.BucketName;
```
Comment (3ti65) [inline VUModManagerRegistry/Services/S3/S3ModStorage.cs:29]: whats this IOptions thingy and why do you need that?

File: VUModManagerRegistry/Services/S3/S3ModStorage.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                Key = CreateModVersionKey(modName, modVersion)
+            });
+        }
+
+        public string GetArchiveDownloadLink(string modName, string modVersion)
+        {
+            return _s3Client.GetPreSignedURL(new GetPreSignedUrlRequest
+            {
+                BucketName = _bucketName,
+                Expires = DateTime.UtcNow.AddSeconds(30),
```
Comment (3ti65) [inline VUModManagerRegistry/Services/S3/S3ModStorage.cs:58]: it makes sense to use a systemTimeProvider class for handling all date / time related things and inject that into your service classes.

**Review by 3ti65 — APPROVED** [review]
See comments

File: VUModManagerRegistry/Services/S3/S3ModStorage.cs (hunk 4 above)
Comment (paulhobbel) [reply VUModManagerRegistry/Services/S3/S3ModStorage.cs:29]: It's something from ASP.net to inject options from your settings file

**Pull request merged**